### ISSUE TITLE
Reconcile `ClusterResourceSet`s

### ISSUE THREAD

**Issue by g-gaston** [kubernetes-sigs/cluster-api#5555]
**User Story**

As an operator I would like to be able to update objects managed by a `ClusterResourceSet` by interacting exclusively with the management cluster for security/stability/new-capabilities updates.

**Detailed Description**

As mentioned in the [design proposal](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200220-cluster-resource-set.md), `ClusterResourceSet`'s only support the `ApplyOnce` mode. This makes impossible to update such resources without interacting with the workload clusters directly. It makes cluster maintenance a bit more cumbersome, since objects need to be reapplied on each cluster individually, as opposed to letting cluster-api manage that complexity. Plus it doesn't guarantee that all workload clusters have the same version of such objects.

I believe the solution to this issue (as is kind of explained in the original CAEP) comes by implementing new mode/s. There are multiple ways to do this: reapply periodically, when the resources hash changes... I'm more than happy to implement this feature but I would love some guidance before starting the work, specially if this is a change that would be welcomed and also which mode/modes should we aim for.

**Anything else you would like to add:**

The VSphere provider right now uses "by default" a `ClusterResourceSet` for the CPI and CSI. So I believe that when having VSphere production clusters, `ClusterResourceSet`, even if still experimental, is not necessarily a nice to have feature anymore, but a key component that would benefit from a better lifecycle management.

Another note: I was expecting, as workaround for this issue, that resources would be reapplied when creating a new `ClusterResourceSet` pointing to the same objects (since this CRS wouldn't have a `ClusterResourceSetBinding`). However, that doesn't happen because of how [this](https://github.com/kubernetes-sigs/cluster-api/blob/main/exp/addons/controllers/clusterresourceset_helpers.go#L92) is implemented. I believe the CAEP doesn't specify that objects won't be reapplied if a new CRS is created, but it also doesn't specify the opposite. Is this maybe something you might be open to change while we work on a new mode?

/kind feature

**Comment by sbueringer** [comment]
@g-gaston Can you please check if this issue is already covered by https://github.com/kubernetes-sigs/cluster-api/issues/4807?

**Comment by vincepri** [comment]
/milestone Next

Sounds like a duplicate of the above, could we merge?

**Comment by g-gaston** [comment]
@sbueringer @vincepri 
Sounds good, let's merge it

**Comment by vincepri** [comment]
/close

@g-gaston Are you able to post your original issue in the other one?

**Comment by g-gaston** [comment]
@vincepri 
Of course. Do you mean the whole issue body? Or just the part about capv?
